### ISSUE TITLE
Class Indices for Unseen Classes

### ISSUE THREAD

**Issue by Yang-Li-2000** [Jun-CEN/Open_world_3D_semantic_segmentation#13]
In the repo "semantic_kitti_api-master", it seems that the class index for unseen classes is 5 after mapping, which corresponds to class 13, 16, 20, 99, 256, 257, and 259 before mapping. 
<img width="713" alt="image" src="https://user-images.githubusercontent.com/50163745/200537166-6b5d2d80-692e-49f9-b910-1c101b85b768.png">
<img width="713" alt="Screen Shot 2022-11-08 at 17 03 57" src="https://user-images.githubusercontent.com/50163745/200537479-d35f7fb7-d968-4c30-bf85-8e787bf0f1f0.png">

In val_cylinder_asym_ood.py in the repo "Open_world_3D_semantic_segmentation", in contrast, there does not seem to be a class for unseen objects. It looks like a few classifiers (self.logits2 in forward_dummy_final in segmentator_3d_asymm_spconv.py) are specialized for producing uncertainty scores for unseen classes. However, it is not clear which class corresponds to the unseen class. 
<img width="713" alt="image" src="https://user-images.githubusercontent.com/50163745/200538489-0f08b47e-ddd2-4486-904e-39b4173b1873.png">

Additionally, in Section 5.1 of the paper, novel classes are defined to be {barrier, construction-vehicle, tra ffi c-cone, trailer}. 
<img width="941" alt="image" src="https://user-images.githubusercontent.com/50163745/200539557-80b468b0-5a49-472f-8990-f373933c103e.png">

In semantic-kitti.yaml in both of the repos, however, there does not seem to exist barrier, construction-vehicle, tra ffi c-cone, or trailer.

I am confused about how to tell which points belong to the unseen class when running val_cylinder_asym_ood.py.

**Comment by Jun-CEN** [comment]
Hi,

(1) The val_cylinder_asym_ood.py. is to generate  and store the uncertainty score for every point. After that, you should use our [semantic_kitti_api](https://github.com/Jun-CEN/semantic_kitti_api) or [nuScenes_api](https://github.com/Jun-CEN/nuScenes_api) to evaluate the open-set performance based on the stored uncertainty files. The unknown class for open-set evaluation is at 
https://github.com/Jun-CEN/semantic_kitti_api/blob/bd918e4b3f143d1f12132b5b5e586abf5f0c7149/auxiliary/np_ioueval.py#L86-L87
(2) K^{sk} is the unknown class for SemanticKITTI dataset, and K^{ns} is the unknown class for nuScenes dataset. So for SemanticKITTI, the unknown class is only other-vehicle.

**Comment by Yang-Li-2000** [comment]
I am still confused about the class indices for unseen objects when using semantic kitti dataset. What are the differences between class 5 and class 20 in voxel_label_origin in train_cylinder_asym_ood_final.py?

**Comment by Jun-CEN** [comment]
Hi,

The unknown class in SemanticKITTI is only 5. The 20 in train_cylinder_asym_ood_final.py is the position of our dummy classifier. In other words, the probability of unknown class is at the 20th position of the logits. loss_dummy means the second largest probability should be the unknown class, which refers to the model calibration technique in the paper.
